### ISSUE TITLE
`skbio.TreeNode` fails to read 1 root branch length for some GreenGenes trees

### ISSUE THREAD

**Issue by stevendbrown** [biocore/scikit-bio#1520]
Trying to compute a weighted UniFrac distance using `beta_diversity` and the 97% GreenGenes tree, I get the following error:

`ValueError: All non-root nodes in ``tree`` must have a branch length.`

I observed this while processing an experiment dataset but it can be reproduced with the following:

```
import glob
import numpy as np
import skbio
from skbio import TreeNode
from skbio.diversity import beta_diversity

all_tree_fns = sorted(glob.glob('/Pynchon/gg_13_8_otus/trees/*.tree'))

for tfn in all_tree_fns:
    tree = TreeNode.read(tfn)
    try:
        wu_dm = beta_diversity("weighted_unifrac", 
                       np.random.randint(low=1, high=20, size=(10, 20)), 
                       [chr(i) for i in range(65, 75)],
                       tree=tree,
                       otu_ids=[y.name for y in list(tree.tips())[:20]])
        print('ok tree:', tfn.split('/')[-1])
    except:
        print('failed tree:', tfn.split('/')[-1])
```

On my computer, this returns the following:

```
ok tree: 61_otus.tree
ok tree: 61_otus_unannotated.tree
ok tree: 64_otus.tree
ok tree: 64_otus_unannotated.tree
ok tree: 67_otus.tree
ok tree: 67_otus_unannotated.tree
failed tree: 70_otus.tree
failed tree: 70_otus_unannotated.tree
failed tree: 73_otus.tree
failed tree: 73_otus_unannotated.tree
failed tree: 76_otus.tree
failed tree: 76_otus_unannotated.tree
failed tree: 79_otus.tree
failed tree: 79_otus_unannotated.tree
failed tree: 82_otus.tree
failed tree: 82_otus_unannotated.tree
failed tree: 85_otus.tree
failed tree: 85_otus_unannotated.tree
failed tree: 88_otus.tree
failed tree: 88_otus_unannotated.tree
failed tree: 91_otus.tree
failed tree: 91_otus_unannotated.tree
failed tree: 94_otus.tree
failed tree: 94_otus_unannotated.tree
failed tree: 97_otus.tree
failed tree: 97_otus_unannotated.tree
failed tree: 99_otus.tree
failed tree: 99_otus_unannotated.tree
```

However, inserting the following two lines after loading the tree resolves the problem:

```
tree.children[1].length = 0.01
tree.children[0].length = 0.01
```

Which branch length is missing depends on which tree is being loaded; setting the branch length of `children[1]` resolves the problem for the >= 76% trees, while setting the branch length of `children[0]` resolves the problem for the 70% and 73% trees.

Trees load without obvious issue using PyCogent in Python 2.

I haven't tried to find the origin of the problem in `skbio` yet.

**Comment by mortonjt** [comment]
The issue is happening because the branch lengths are not assigned for some of the edges in the greengenes tree.  The quick fix around this is to run the following

```
for n in tree.postorder(include_self=True):
    if n.length is None:
        n.length = 0
```
That will force all of the branch lengths to have some integer valued number.  I don't think this is a problem with beta diversity, but rather the input tree.  Maybe worthwhile adding a preprocessing function to mitigate these sorts of scenarios?

**Comment by wasade** [comment]
Correct. It's only a single edge and stems from the reroot performed on the construction of the Greengenes tree where the added node did not get a branch length set. We didn't recognize it at the time as PyCogent assumed zero branch length in the case of an edge without an explicit length.

@mortonjt, I think this is the correct behavior as skbio raises an appropriate error about the tree. Defaulting to assuming zero may lead to unexpected results for a user, whereas if the error is exposed as is right now, then the user is aware of a problem with the tree.

**Comment by mortonjt** [comment]
Exactly - I don't think we should change `beta_diversity`.  But maybe we should have a preprocessing step outside of `beta_diversity`, so that the user can easily correct this problem (i.e. some analogous to the preprocessing modules in sklearn).

**Comment by stevendbrown** [comment]
Would it be reasonable to add something to the docs which describes this behavior and it's interaction with trees, maybe something to the effect of @mortonjt's initial reply (including the code snippet) to the `beta_diversity` or `TreeNode` docs? I suppose if I'm the only one who has ever noticed this, it's probably not worth a bunch of hacking to implement a change, but it might help to avoid confusion for new users?

Regardless of all that, thank you both @mortonjt and @wasade for your comments, I'm definitely learning a lot here! I'm happy to ponder what could be added to docs and how and submit a PR when I think I have something good.
